### ISSUE TITLE
How best can I propose some icon fixes?

### ISSUE THREAD

**Issue by scottjenson** [mastodon/mastodon#19513]
### Pitch

I was hoping to just inspect an icon, make a change and do a pull request (with appropriate discussion of course) However, the code is a bit hard to follow. It appears many icons are done through :before elements and JS so I can't easily see the name or the directory of the icon in question. I'd like to be of help to this project but if I can't browse the assets easily, it's a bit of a challenge.

The icon in question is the 'bell' icon that shows up on a profile page. I can see it has a CSS style "fa-bell" but I can't really get past that. Looking for just a bit of orientation.

### Motivation

Two separate points: 1) fix the immediate issue about the "notification icon" has two states that are visually identical. Just making them more distinct (yet still in the same style) would be a simple, quick UX win. However, the bigger issue #2 is for there to be a bit more documentation so UX people (who don't always have strong sw skills) can still be helpful.

I understand these things need to be discussed. I'm not asking for a decision to be made here. I'm just asking for a bit of help so UX designers can help out. Let's at least unlock their ability to contribute. I'm looking for a tiny bit of guidance on where the graphical assets are stored for mastodon.

**Comment by Gargron** [comment]
In the web app we still don't have custom icons, we are just using Font Awesome v4.

**Comment by scottjenson** [comment]
That's fine, I would still expect that Font Awesome would offer better options to pick from. The hard part is finding the code that makes these choices. Is there a general rule to find the code that makes this choice?

For example, the choice I'm suggested specifically is that you change the code from using bell-o
https://github.com/mastodon/mastodon/issues/19513

to using just normal bell instead as it makes for a stronger contrast
https://github.com/mastodon/mastodon/issues/19513

Let's be clear, this is a TINY change. I'm just using it as a 'gateway' issue so I can learn a bit more about the code and how I can offer PRs to the team.

**Comment by Gargron** [comment]
There is most likely a `<Icon id='bell' />` in the JS code.

**Comment by scottjenson** [comment]
Ok, I believe I've found the line that needs to change:
app/javascript/mastodon/features/account/components/header.js line 172
change "bell-o" to "bell"

However, as this is react there is no way I can build this and test it out. I can't even use Chrome DevTools to edit the class name.

I'd like to help but even this incredibly trivial change requires a core member of the team. I totally understand if you'd like to skip this. It's just too bad it's so incredibly complex to make such a trivial change.

**Comment by ClearlyClaire** [comment]
> Ok, I believe I've found the line that needs to change: app/javascript/mastodon/features/account/components/header.js line 172 change "bell-o" to "bell"

This suggestion has already been made to @Gargron repeatedly, he was opposed to that at the time but that was a while ago, so he may have changed his mind in the meantime, though.

> However, as this is react there is no way I can build this and test it out.

You can definitely build it, although it is pretty involved. If you are interested, you could look at https://docs.joinmastodon.org/dev/setup/ though I am not completely sure how up to date it is.

> I can't even use Chrome DevTools to edit the class name.

You should be able to, although the change won't persist once the component gets completely re-rendered, but that should not happen unless you leave the profile or follow/unfollow the person. Not ideal, but it should do for quick testing.

**Comment by scottjenson** [comment]
Thanks. Helpful to have the history. Given how nontrivial it is to build, I'm unlikely to go that direction. It would be helpful if that wasn't the only path to helping. I've reached out to @lickability as they (apparently) do most of the UX work? If so I'm hoping they can help coordinate some of this work.

As to "viewing in Chrome DevTools", normally with an <img> tag it's trivial to just change src. However, with React, its done through a bizantine set of CSS classes (and :before pseudo classes) so it's not at all easy to see how to change the icon from "bell-o" to "bell" (at least, not easy for me)

**Comment by ClearlyClaire** [comment]
Lickability did the UX work on the mobile apps, which in turn has been reflected in parts of the WebUI, but they weren't there from the start nor have they directly worked on the WebUI as far as I know. I don't think that feature's design went through them.

> As to "viewing in Chrome DevTools", normally with an img tag it's trivial to just change src. However, with React, its done through a bizantine set of CSS classes (and :before pseudo classes) so it's not at all easy to see how to change the icon from "bell-o" to "bell" (at least, not easy for me)

React itself is pretty transparent, it's just our code having a few layers of indirection. An `<IconButton>` component such as the one you're looking at translates to a `button` tag with a `i` tag in it. The `i` tag in this case has a `fa-bell-o` class that is the one you want to change. The `:before` pseudo-class thing is (one of) FontAwesome's way of working and you don't need to be concerned about it.

**Comment by Gargron** [comment]
Worth saying, I would *love* to have custom icons instead of Font Awesome at some point.

**Comment by scottjenson** [comment]
Especially if you used SVG so you didn't have to worry about resolution issues (at last on the vast majority of devices) One possibility is to pick a tiny subsystem and convert just that and see how it goes. Let me know if you'd like help on that.

**Comment by scottjenson** [comment]
@ClearlyClaire Damn! That worked perfectly! Thank you. This at least allows me to reasonably test changes (with different states) so when I make the PR there is some visual confirmation that it's reasonable.

**Comment by nhoizey** [comment]
> Especially if you used SVG 

Related: #1469

**Comment by nclm** [comment]
Just for fun (I had to try!), Mastodon with the [Streamline Flex](https://www.streamlinehq.com/icons/streamline-flex---free) icons previously mentioned:

![MastodonDark_Icons_StreamlineFlex_1](https://user-images.githubusercontent.com/923008/199820684-9333ab73-c190-43a6-8649-f35f27a0b648.png)

Not all of the icons here are exactly matching the needs of course, but as a proof of concept for another icon set, in case anyone wasn’t convinced, it shows how much changing the icon design can transform the general mood of the interface: in that case, see how much friendlier Mastodon looks with these!

**Comment by Dehelssey** [comment]
Looks **so much better**

**Comment by scottjenson** [comment]
It does look better! Is this a visual (e.g. Figma) mockup? The next step would be to change the CSS to reference these icons. Given how React works, I doubt this would be a trivial change. 

I also agree with a previous comment that if we move in this direction, there should be an 'icon library' that every client can use. Not for 'extreme compatibility' but simple convenience. *Nobody* wants to build their own icon set but if there is a default set lying around... it's so much easier

**Comment by Gargron** [comment]
I believe we have some icons that we use in both the iOS and the Android app, that we would be most likely to adopt in the web app as well (I have to ask @samhenrigold about that). The big challenge is that the web app uses *a lot* of different icons.

**Comment by katossky** [comment]
I am not sure whether I should open a new discussion, but may I suggest to update the follow/no-follow icons? At the moment they only differ slightly (+ for follow, x for no follow) but look the same at first glance. This is annoying when searching for persons from the search bar, as it happened several times that I was unfollowing someone I was already subscribed to. Just making the no-follow icon red would help.

**Comment by nclm** [comment]
> Is this a visual (e.g. Figma) mockup? 

It is, well, kind of. It’s a very very quick and dirty job pasting small screenshots of the icons over a screenshot of Mastodon in Gimp. That said, it would great to recreate the Mastodon interface in Penpot to enable designers (and some non-designers) who don’t code to easily create suggestions and fixes for the UI…

> I believe we have some icons that we use in both the iOS and the Android app, that we would be most likely to adopt in the web app as well (I have to ask @samhenrigold about that). 

Ah great to hear, would be good also to have consistency between apps and web! 

> The big challenge is that the web app uses a lot of different icons.

Yes, it would actually be interesting to list them all and their function. We could also think in two phases, first the icons that appears in the actual social networking UI, and then all the icons in the preferences, moderation and admin panels, which are not as essential.

(Sorry this is getting a little out of topic with the original issue. Should we create a specific “custom icon theme” issue? Can we move messages from an issue to another?)

**Comment by robsonsobral** [comment]
Could we, please, use https://fontawesome.com/v4/icon/bell-slash and https://fontawesome.com/v4/icon/bell as a quick fix?

Until all the bells and whistles get ready, this can't get more confusing.

**Comment by scottjenson** [comment]
Agreed, this is a trivial change that would have a large positive UX impact. It was said above that @Gargron is against this change? Is that still true?

**Comment by vmstan** [comment]
SVG icon change has been merged to main https://github.com/mastodon/mastodon/pull/27385

**Comment by renchap** [comment]
Closing this as we recently changed the icons to use the Material Icons.
